src/charm.py

```diff
@@ -276,6 +280,27 @@ def _check_grpc(self, interfaces):
             raise CheckFailed("Waiting for upstream gRPC connection information.", WaitingStatus)
         return upstreams
 
+    def _send_data_to_ingress_provider(self, interfaces):
+        """Send data to the ingress relation data bag so the VirtualServices provider configures
+        a VirtualService routing traffic from `/ml_metadata` path to envoy service.
+
+        Raises an exception and sets the charm to Blocked if there is no ingress relation available
+        """
+        if interfaces["ingress"]:
+            interfaces["ingress"].send_data(
+                {
+                    "prefix": "/ml_metadata",
+                    "rewrite": "/ml_metadata",
+                    "service": self.model.app.name,
+                    "port": int(self.model.config["http-port"]),
+                }
+            )
+        else:
+            raise CheckFailed(
+                "No 'ingress' relation available, please relate to a VirtualServices provider e.g. istio-pilot.",  # noqa E501
```

Comment:
Usually these messages get cut by the juju status output. Maybe re-phrase it so it's shorter: "Please relate to istio-pilot, no 'ingress' relation available' or similar.

Reply:
pushed a commit for this